### PR TITLE
Fix remaining blackmail vulnerabilities

### PR DESCRIPTION
**Pull request by stejbac**
This PR attempts to take advantage of the SegWit trade protocol upgrade (hard fork), to fix all the remaining cases where a buyer or seller can publish a deposit tx without their peer having a valid delayed payout tx (at least in the fully SegWit case, when taking an offer created in v1.5.0+). When taking pre-1.5.0 offers, the deposit tx inputs will be of mixed type, which makes it malleable and thus impossible to completely prevent either trader from altering its ID and thus rendering their peer's delayed payout tx invalid. In that case, the buyer should still fully validate their delayed payout tx anyway, against the agreed deposit tx, so that if the latter changes before confirming, the trade hopefully fails before the buyer makes payment. The PR additionally fixes a bug where (it appears) the buyer fails to check the signature of the final delayed payout tx.

These changes are intended to prevent blackmail attacks (or reduce the risk of them in not-fully-SegWit cases where the deposit tx is malleable).

### CODE REVIEW

File: core/src/main/java/bisq/core/trade/protocol/tasks/buyer_as_maker/BuyerAsMakerSendsInputsForDepositTxResponse.java
Comment (chimp1984) [inline core/src/main/java/bisq/core/trade/protocol/tasks/buyer_as_maker/BuyerAsMakerSendsInputsForDepositTxResponse.java:40]: The Transaction.bitcoinSerializeToStream method does use that flag not only for adding segwit data but also for writing a flag: 
```
 // marker, flag
        if (useSegwit) {
            stream.write(0);
            stream.write(1);
        }
```

I don't know which function this flag has, but maybe its more safe to remove the segwit data in a different way. Maybe @oscarguindzberg can add his input here...

**Review by chimp1984 — CHANGES_REQUESTED** [review]

File: core/src/main/java/bisq/core/trade/protocol/tasks/buyer_as_maker/BuyerAsMakerSendsInputsForDepositTxResponse.java
Comment (oscarguindzberg) [reply core/src/main/java/bisq/core/trade/protocol/tasks/buyer_as_maker/BuyerAsMakerSendsInputsForDepositTxResponse.java:40]: the function of the flag is "tx serialzed using the new segwit format".
bitcoinSerialize(false) seems fine if you want to avoid sending the signatures.

**Comment by sqrrm** [comment]
I like this PR and it should be included in v1.5.0. From what I can see it looks reasonable but I'm not familiar enough with all the bitcoinj details.

**Comment by stejbac** [comment]
I just spotted that I accidentally used the non-segwit version of `Script.correctlySpends`, when checking the p2wsh signature of the delayed payout tx (during `BuyerFinalizesDelayedPayoutTx`), on line 774 of `TradeWalletService`:

> input.getScriptSig().correctlySpends(delayedPayoutTx, 0, scriptPubKey, Script.ALL_VERIFY_FLAGS);

This erroneously skips the check, as the ScriptSig is empty. I need to additionally pass the witness and locked up trade amount. I'll post another commit to fix this shortly.

**Review by chimp1984 — APPROVED** [review]
<strike>utACK</strike>

Ah read the last comment just now. So I will wait for the open fix for another review...

**Comment by chimp1984** [comment]
@oscarguindzberg Could you please review that PR as well?

**Review by chimp1984 — APPROVED** [review]
utACK

**Comment by sqrrm** [comment]
@stejbac The build is failing

@oscarguindzberg I will wait for you ack before merging this

**Comment by stejbac** [comment]
@sqrrm I'm not sure why the build is failing (or the Codacy checks). The Travis build log of one job, which ran 9 hours ago, has the following error:

> Could not GET 'https://repo.maven.apache.org/maven2/org/openjfx/javafx-base/11.0.2/javafx-base-11.0.2.jar'.

The other job (apparently run around the same time) has failures in `OpenOfferManagerTest` (lines 45, 106) in the _:core:test_ task, but I'm fairly sure that's not caused by my changes.

**Comment by ghubstan** [comment]
> ... I'm not sure why the build is failing (or the Codacy checks). The Travis build log of one job, which ran 9 hours ago, has the following error:
> 
> > Could not GET 'https://repo.maven.apache.org/maven2/org/openjfx/javafx-base/11.0.2/javafx-base-11.0.2.jar'.
> 
> The other job (apparently run around the same time) has failures in `OpenOfferManagerTest` (lines 45, 106) in the _:core:test_ task, but I'm fairly sure that's not caused by my changes.

Which jar travis cannot download at any given time is fairly random.  Closing and re-opening the PR will re-run the travis build, and the dependency download will probably work.    However, Codacy's `Seems like the release/v1.5.0 is not being analyzed. To review this pull request enable the branch in the repository settings.` is another issue that someone (@ripcurlx ?) needs to look at.

**Comment by ripcurlx** [comment]
> > ... I'm not sure why the build is failing (or the Codacy checks). The Travis build log of one job, which ran 9 hours ago, has the following error:
> > > Could not GET 'https://repo.maven.apache.org/maven2/org/openjfx/javafx-base/11.0.2/javafx-base-11.0.2.jar'.
> > 
> > 
> > The other job (apparently run around the same time) has failures in `OpenOfferManagerTest` (lines 45, 106) in the _:core:test_ task, but I'm fairly sure that's not caused by my changes.
> 
> Which jar travis cannot download at any given time is fairly random. Closing and re-opening the PR will re-run the travis build, and the dependency download will probably work. However, Codacy's `Seems like the release/v1.5.0 is not being analyzed. To review this pull request enable the branch in the repository settings.` is another issue that someone (@ripcurlx ?) needs to look at.

Yes, please ignore the Codacy warning. It is just about the fact that it is not configured to run on this branch. I just manually triggered a travis re-build for this PR.

**Comment by ripcurlx** [comment]
I just saw in Travis that there are two failing tests right now:
```
bisq.core.offer.OpenOfferManagerTest > testStartEditOfferForActiveOffer FAILED
    java.lang.RuntimeException at OpenOfferManagerTest.java:45
bisq.core.offer.OpenOfferManagerTest > testStartEditOfferForOfferThatIsCurrentlyEdited FAILED
    java.lang.RuntimeException at OpenOfferManagerTest.java:106
```

**Comment by ripcurlx** [comment]
> I just saw in Travis that there are two failing tests right now:
> 
> ```
> bisq.core.offer.OpenOfferManagerTest > testStartEditOfferForActiveOffer FAILED
>     java.lang.RuntimeException at OpenOfferManagerTest.java:45
> bisq.core.offer.OpenOfferManagerTest > testStartEditOfferForOfferThatIsCurrentlyEdited FAILED
>     java.lang.RuntimeException at OpenOfferManagerTest.java:106
> ```

🤔 it does run locally without any problems.

**Comment by stejbac** [comment]
I got the failing tests locally after I tried merging onto release/1.5.0, so the tip map be broken right now. It seems it can be fixed by adding tearDown() methods in the failing tests to call PersistenceManager.shutDown, which removes the particular persistence manager instance created in each test from a global map. (All but one tests in each affected test class fail, presumably depending in the order they run in.)

**Comment by ripcurlx** [comment]
> I got the failing tests locally after I tried merging onto release/1.5.0, so the tip map be broken right now. It seems it can be fixed by adding tearDown() methods in the failing tests to call PersistenceManager.shutDown, which removes the particular persistence manager instance created in each test from a global map. (All but one tests in each affected test class fail, presumably depending in the order they run in.)

Yes, also https://github.com/bisq-network/bisq/pull/4816 is affected by this right now. Could you please push this commit to this branch to see if Travis is happy with it as well and I'll cherry-pick and push it to the other one as well?

**Comment by stejbac** [comment]
It looks like it fixed the jdk10 job but not the jdk12 job, which I guess has an unrelated problem.

**Comment by ripcurlx** [comment]
> It looks like it fixed the jdk10 job but not the jdk12 job, which I guess has an unrelated problem.

I just restarted the job. Travis is kind of unstable the last couple of days.

**Comment by ripcurlx** [comment]
@stejbac As I just merged #4816. Could you please resolve the conflict? Thanks!

**Comment by stejbac** [comment]
@ripcurlx OK, I guess I should rebase onto release/1.5.0 to resolve the conflict.

**Review by chimp1984 — APPROVED** [review]
utACK

**Pull request merged**
